src/API/QueryFieldService.php

```diff
@@ -0,0 +1,103 @@
+use eZ\Publish\Core\QueryType\QueryTypeRegistry;
+use Symfony\Component\ExpressionLanguage\ExpressionLanguage;
+use Symfony\Component\Yaml\Yaml;
+
+/**
+ * Executes a query and returns the results.
+ */
+class QueryFieldService
+{
+    /** @var \eZ\Publish\Core\QueryType\QueryTypeRegistry */
+    private $queryTypeRegistry;
+
+    /** @var \eZ\Publish\API\Repository\SearchService */
+    private $searchService;
+
+    /** @var \eZ\Publish\API\Repository\ContentTypeService */
+    private $contentTypeService;
+
+    public function __construct(
+        SearchService $searchService,
+        ContentTypeService $contentTypeService,
+        QueryTypeRegistry $queryTypeRegistry
+    ) {
+        $this->searchService = $searchService;
+        $this->contentTypeService = $contentTypeService;
+        $this->queryTypeRegistry = $queryTypeRegistry;
+    }
+
+    /**
+     * @param \eZ\Publish\API\Repository\Values\Content\Content $content
+     * @param string $fieldDefinitionIdentifier
+     *
+     * @return \eZ\Publish\API\Repository\Values\Content\Content[]
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
+     */
+    public function loadFieldData(Content $content, string $fieldDefinitionIdentifier): array
+    {
+        $fieldDefinition = $this->getFieldDefinition($content->contentInfo, $fieldDefinitionIdentifier);
+        $queryType = $this->queryTypeRegistry->getQueryType($fieldDefinition->fieldSettings['QueryType']);
+        $parameters = $this->resolveParameters($fieldDefinition->fieldSettings['Parameters'], $content);
+
+        return array_map(
+            function (SearchHit $searchHit) {
+                return $searchHit->valueObject;
+            },
+            $this->searchService->findContent($queryType->getQuery($parameters))->searchHits
+        );
+    }
+
+    public function getFieldDefinition(ContentInfo $contentInfo, string $fieldDefinitionIdentifier): FieldDefinition
+    {
+        return $queryFieldDefinition =
+            $this
+                ->contentTypeService->loadContentType($contentInfo->contentTypeId)
+                ->getFieldDefinition($fieldDefinitionIdentifier);
+    }
+
+    /**
+     * @param string $parameters parameters in JSON format
+     * @param \eZ\Publish\API\Repository\Values\Content\Content $content
+     *
+     * @return array
+     */
+    private function resolveParameters(string $parametersString, Content $content): array
+    {
+        $parameters = [];
+        foreach (Yaml::parse($parametersString) as $key => $parameter) {
+            $parameters[$key] = $this->applyContentToParameter($content, $parameter);
+        }
+
+        return $parameters;
+    }
+
+    protected function applyContentToParameter(Content $content, string $parameter)
+    {
+        if (substr($parameter, 0, 2) !== '@=') {
+            return $parameter;
+        }
+
+        return (new ExpressionLanguage())->evaluate(
```

Comment:
Q: What about error handling?

Reply:
Follow-up ? :)

Reply:
This is actually a very interesting question. How can we validate the expression ? We don't have a content we can run at our disposal. Should we mock an item ? It sounds a bit tricky... it could be a good feature, though (*Gimme a fake Article content item for a read-only test*) :-)